### ISSUE TITLE
Potential deadlock in pkg/monitor

### ISSUE THREAD

**Issue by tgraf** [cilium/cilium#10258]
Observed in https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/17327/testReport/(root)/Suite-runtime/RuntimeLB_With_Containers_validates_service_recovery_on_restart/

```
Feb 20 02:28:30 runtime cilium-agent[9814]: level=debug msg="Upserting service entry" slaveSlot=1 subsys=map-lb svcKey="127.0.0.1:80" svcVal="3 (1) [FLAGS: 0x0]"
Feb 20 02:28:30 runtime cilium-agent[9814]: level=debug msg="Upserting service entry" slaveSlot=2 subsys=map-lb svcKey="127.0.0.1:80" svcVal="4 (1) [FLAGS: 0x0]"
Feb 20 02:28:30 runtime cilium-agent[9814]: level=debug msg="Upserting service entry" slaveSlot=0 subsys=map-lb svcKey="127.0.0.1:80" svcVal="0 (1) [FLAGS: 0x0]"
Feb 20 02:28:30 runtime cilium-agent[9814]: level=debug msg="GET /service/{id} request" params="{HTTPRequest:0xc001fb3700 ID:20}" subsys=daemon
Feb 20 02:28:30 runtime cilium-agent[9814]: level=debug msg="Listener disconnected" subsys=monitor-agent
Feb 20 02:28:30 runtime cilium-agent[9814]: POTENTIAL DEADLOCK: Duplicate locking, saw callers this locks in one goroutine:
Feb 20 02:28:30 runtime cilium-agent[9814]: current goroutine 153205 lock &{{9 0}}
Feb 20 02:28:30 runtime cilium-agent[9814]: all callers to this lock in the goroutine
Feb 20 02:28:30 runtime cilium-agent[9814]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/lock/lock_debug.go:91 lock.(*internalMutex).Lock { i.Mutex.Lock() } <<<<<
Feb 20 02:28:30 runtime cilium-agent[9814]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/monitor/agent/monitor.go:208 agent.(*Monitor).handleEvents { m.Lock() }
Feb 20 02:28:30 runtime cilium-agent[9814]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/lock/lock_debug.go:91 lock.(*internalMutex).Lock { i.Mutex.Lock() } <<<<<
Feb 20 02:28:30 runtime cilium-agent[9814]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/monitor/agent/monitor.go:180 agent.(*Monitor).handleEvents.func1 { m.Lock() }
Feb 20 02:28:30 runtime cilium-agent[9814]: /usr/local/go/src/runtime/panic.go:679 runtime.gopanic { reflectcall(nil, unsafe.Pointer(d.fn), deferArgs(d), uint32(d.siz), uint32(d.siz)) }
Feb 20 02:28:30 runtime cilium-agent[9814]: /usr/local/go/src/runtime/chan.go:187 runtime.chansend { panic(plainError("send on closed channel")) }
Feb 20 02:28:30 runtime cilium-agent[9814]: /usr/local/go/src/runtime/chan.go:615 runtime.selectnbsend { return chansend(c, elem, false, getcallerpc()) }
Feb 20 02:28:30 runtime cilium-agent[9814]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/monitor/agent/listener1_2.go:51 agent.(*listenerv1_2).Enqueue { case ml.queue <- pl: }
Feb 20 02:28:30 runtime cilium-agent[9814]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/monitor/agent/monitor.go:273 agent.(*Monitor).sendLocked { ml.Enqueue(pl) }
Feb 20 02:28:30 runtime cilium-agent[9814]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/monitor/agent/monitor.go:220 agent.(*Monitor).handleEvents { m.sendLocked(&pl) }
```

**Comment by joestringer** [comment]
I suspect that the interesting element here is the panic for `send on closed channel`.

**Comment by joestringer** [comment]
Commit 926ad33a7510c88b3a0ad6d31b3328127e1a57a1 is a potentially relevant recent change, though I'm surprised it would take a month before we noticed in CI.

**Comment by joestringer** [comment]
@michi-covalent as discussed out-of-band, you'll take a look. If you don't have cycles or reach a dead-end, feel free to update this issue and un-assign.

**Comment by michi-covalent** [comment]
i guess the easiest "fix" is:
- explicitly close the channel here: https://github.com/cilium/cilium/blob/a1d93e044c1f2064da008a80ffe593bd83a9a669/pkg/monitor/agent/listener1_2_test.go#L45
- don't close the channel here: https://github.com/cilium/cilium/blob/a1d93e044c1f2064da008a80ffe593bd83a9a669/pkg/monitor/agent/listener1_2.go#L89 let it get garbage collected instead. the close was added for the unit test above so that the drainQueue gets unblocked and the cleanup function gets called.

**Comment by michi-covalent** [comment]
looking into this a bit more, i'm not sure if it's related to closing the queue inside the listener.

the log message is saying that this Lock() gets called:

```
Feb 20 02:28:30 runtime cilium-agent[9814]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/lock/lock_debug.go:91 lock.(*internalMutex).Lock { i.Mutex.Lock() } <<<<<
Feb 20 02:28:30 runtime cilium-agent[9814]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/monitor/agent/monitor.go:180 agent.(*Monitor).handleEvents.func1 { m.Lock() }
```
before this once gets Unlock()ed:
```
Feb 20 02:28:30 runtime cilium-agent[9814]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/lock/lock_debug.go:91 lock.(*internalMutex).Lock { i.Mutex.Lock() } <<<<<
Feb 20 02:28:30 runtime cilium-agent[9814]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/monitor/agent/monitor.go:208 agent.(*Monitor).handleEvents { m.Lock() }
```

looking at handleEvents https://github.com/cilium/cilium/blob/a1d93e044c1f2064da008a80ffe593bd83a9a669/pkg/monitor/agent/monitor.go#L164 i'm not sure how that's possible. 

one thing i can try is to upgrade the go-deadlock version. the vendored version is quite a bit behind the latest (https://github.com/sasha-s/go-deadlock/tree/v0.2.0). there might have been some fixes.

let me know if you have any other suggestions.

**Comment by joestringer** [comment]
I agree that that's what the deadlock message is saying, and with a cursory glance I don't have an argument as to why that might be possible either.

However, I think the more interesting aspect of this report is the panic for send on closed channel. Just after this log is printed, the daemon shuts down and restarts which suggests a real failure case.

Come to think of it, the deadlock could be caused by the panic path not unlocking but instead triggering the deferred function, leading to double-unlock.

#10274 has a full zip, but here's a slightly bigger snippet of the cilium-agent logs including the restart:

```
Feb 20 13:20:47 runtime cilium-agent[27474]: level=debug msg="New listener connected" count.listener=2 subsys=monitor-agent version=1.2
Feb 20 13:20:47 runtime cilium-agent[27474]: level=debug msg="Listener disconnected" subsys=monitor-agent
Feb 20 13:20:47 runtime cilium-agent[27474]: POTENTIAL DEADLOCK: Duplicate locking, saw callers this locks in one goroutine:
Feb 20 13:20:47 runtime cilium-agent[27474]: current goroutine 13559 lock &{{9 0}}
Feb 20 13:20:47 runtime cilium-agent[27474]: all callers to this lock in the goroutine
Feb 20 13:20:47 runtime cilium-agent[27474]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/lock/lock_debug.go:91 lock.(*internalMutex).Lock { i.Mutex.Lock() } <<<<<
Feb 20 13:20:47 runtime cilium-agent[27474]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/monitor/agent/monitor.go:208 agent.(*Monitor).handleEvents { m.Lock() }
Feb 20 13:20:47 runtime cilium-agent[27474]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/lock/lock_debug.go:91 lock.(*internalMutex).Lock { i.Mutex.Lock() } <<<<<
Feb 20 13:20:47 runtime cilium-agent[27474]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/monitor/agent/monitor.go:180 agent.(*Monitor).handleEvents.func1 { m.Lock() }
Feb 20 13:20:47 runtime cilium-agent[27474]: /usr/local/go/src/runtime/panic.go:679 runtime.gopanic { reflectcall(nil, unsafe.Pointer(d.fn), deferArgs(d), uint32(d.siz), uint32(d.siz)) }
Feb 20 13:20:47 runtime cilium-agent[27474]: /usr/local/go/src/runtime/chan.go:187 runtime.chansend { panic(plainError("send on closed channel")) }
Feb 20 13:20:47 runtime cilium-agent[27474]: /usr/local/go/src/runtime/chan.go:615 runtime.selectnbsend { return chansend(c, elem, false, getcallerpc()) }
Feb 20 13:20:47 runtime cilium-agent[27474]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/monitor/agent/listener1_2.go:51 agent.(*listenerv1_2).Enqueue { case ml.queue <- pl: }
Feb 20 13:20:47 runtime cilium-agent[27474]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/monitor/agent/monitor.go:273 agent.(*Monitor).sendLocked { ml.Enqueue(pl) }
Feb 20 13:20:47 runtime cilium-agent[27474]: /home/vagrant/go/src/github.com/cilium/cilium/pkg/monitor/agent/monitor.go:220 agent.(*Monitor).handleEvents { m.sendLocked(&pl) }
Feb 20 13:20:47 runtime cilium-agent[27474]: Other goroutines holding locks:
Feb 20 13:20:47 runtime systemd[1]: cilium.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Feb 20 13:20:47 runtime systemd[1]: cilium.service: Failed with result 'exit-code'.
Feb 20 13:20:47 runtime systemd[1]: cilium.service: Service hold-off time over, scheduling restart.
Feb 20 13:20:47 runtime systemd[1]: cilium.service: Scheduled restart job, restart counter is at 1.
Feb 20 13:20:47 runtime systemd[1]: Stopped cilium.
Feb 20 13:20:47 runtime systemd[1]: Started cilium.
```
